language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -58,48 +100,84 @@ type VarEnv = UniqueMap TermUnique Closure
 
 -- | The environment the CEK machine runs in.
 data CekEnv = CekEnv
-    { _cekEnvMeans  :: DynamicBuiltinNameMeanings
-    , _cekEnvVarEnv :: VarEnv
+    { _cekEnvMeans      :: DynamicBuiltinNameMeanings
+    , _cekEnvVarEnv     :: VarEnv
+    , _cekEnvBudgetMode :: CekBudgetMode
     }
 
+makeLenses ''CekEnv
+
 -- | The monad the CEK machine runs in.
-type CekM = ReaderT CekEnv (Either CekEvaluationException)
+type CekM = ReaderT CekEnv (ExceptT CekEvaluationException (State ExBudgetState))
+
+spendBoth :: WithMemory Term -> ExCPU -> ExMemory -> CekM ()
+spendBoth term cpu mem = spendCPU term cpu >> spendMemory term mem
+
+spendBudget
+    :: (Ord a, Num a)
+    => CekBudgetMode
+    -> Lens' ExBudgetState a
+    -> a
+    -> CekM ()
```

Comment:
If you're going to run in `CekM`, you might as well get the `CekBudgetMode` from the env.

Reply:
BTW, I changed this type signature from a more general mtl-ey one (something was hard when I tried to update the code, but keep the signature). I think having simple types is nice anyway.